plugin/output/clickhouse/clickhouse.go

```diff
@@ -0,0 +1,351 @@
+	// >
+	// > A minimum size of events in a batch to send.
+	// > If both batch_size and batch_size_bytes are set, they will work together.
+	BatchSizeBytes  cfg.Expression `json:"batch_size_bytes" default:"0" parse:"expression"` // *
+	BatchSizeBytes_ int
+
+	// > @3@4@5@6
+	// >
+	// > After this timeout batch will be sent even if batch isn't completed.
+	BatchFlushTimeout  cfg.Duration `json:"batch_flush_timeout" default:"200ms" parse:"duration"` // *
+	BatchFlushTimeout_ time.Duration
+}
+
+func init() {
+	fd.DefaultPluginRegistry.RegisterOutput(&pipeline.PluginStaticInfo{
+		Type:    outPluginType,
+		Factory: Factory,
+	})
+}
+
+func Factory() (pipeline.AnyPlugin, pipeline.AnyConfig) {
+	return &Plugin{}, &Config{}
+}
+
+func (p *Plugin) RegisterMetrics(ctl *metric.Ctl) {
+	p.execErrorsMetric = ctl.RegisterCounter("output_clickhouse_errors", "Total clickhouse exec errors")
+	p.writtenEventMetric = ctl.RegisterCounter("output_clickhouse_event_written", "Total events written to ???")
+}
+
+func (p *Plugin) Start(config pipeline.AnyConfig, params *pipeline.OutputPluginParams) {
+	p.logger = params.Logger
+	p.config = config.(*Config)
+	p.ctx, p.cancelFunc = context.WithCancel(context.Background())
+	p.avgEventSize = params.PipelineSettings.AvgEventSize
+
+	if p.config.Retry < 1 {
+		p.logger.Fatal("'retry' can't be <1")
+	}
+	if p.config.Retention_ < 1 {
+		p.logger.Fatal("'retention' can't be <1")
+	}
+	if p.config.DBRequestTimeout_ < 1 {
+		p.logger.Fatal("'db_request_timeout' can't be <1")
+	}
+	if _, err := inferInsaneColInputs(p.config.Schema); err != nil {
+		p.logger.Fatalf("invalid database schema: %s", err)
+	}
+
+	var compression ch.Compression
+	switch strings.ToLower(p.config.Compression) {
+	default:
+		fallthrough
+	case "disabled":
+		compression = ch.CompressionDisabled
+	case "lz4":
+		compression = ch.CompressionLZ4
+	case "zstd":
+		compression = ch.CompressionZSTD
+	case "none":
+		compression = ch.CompressionNone
+	}
+
+	var b tls.ConfigBuilder
+	if p.config.CACert != "" {
+		b := tls.NewConfigBuilder()
+		err := b.AppendCARoot(p.config.CACert)
+		if err != nil {
+			p.logger.Fatalf("can't append CA root: %s", err.Error())
+		}
+	}
+
+	var err error
+	p.pool, err = chpool.Dial(p.ctx, chpool.Options{
+		ClientOptions: ch.Options{
+			Logger:           p.logger.Desugar().Named("driver"),
+			Address:          p.config.Address,
+			Database:         p.config.Database,
+			User:             p.config.User,
+			Password:         p.config.Password,
+			QuotaKey:         p.config.QuotaKey,
+			Compression:      compression,
+			Settings:         p.config.ClickhouseSettings.toProtoSettings(),
+			DialTimeout:      time.Second * 10,
+			TLS:              b.Build(),
+			HandshakeTimeout: time.Minute,
+		},
+		MaxConnLifetime:   time.Minute * 30,
+		MaxConnIdleTime:   time.Minute * 5,
+		MaxConns:          int32(p.config.WorkersCount_),
+		MinConns:          int32(p.config.WorkersCount_),
+		HealthCheckPeriod: time.Minute,
```

Comment:
Could you please move all timeout and period values to plugin config

Reply:
done